### ISSUE TITLE
dependency issue when running scripts/unshard.py

### ISSUE THREAD

**Issue by viking-sudo-rm** [allenai/OLMo#744]
### 🐛 Describe the bug

When I try to run the unshard script on a downloaded checkpoint, I get the following version error related to Pytorch FSDP:

```
Traceback (most recent call last):
  File "/home/willm/OLMo/scripts/unshard.py", line 108, in <module>
    main(
  File "/home/willm/OLMo/scripts/unshard.py", line 36, in main
    model_state_dict, optim_state_dict, trainer_state_dict = checkpointer.unshard_checkpoint(
  File "/home/willm/OLMo/olmo/checkpoint.py", line 1767, in unshard_checkpoint
    model_state = torch.load(path, map_location="cpu")
  File "/home/willm/.local/lib/python3.10/site-packages/torch/serialization.py", line 1025, in load
    return _load(opened_zipfile,
  File "/home/willm/.local/lib/python3.10/site-packages/torch/serialization.py", line 1446, in _load
    result = unpickler.load()
  File "/home/willm/.local/lib/python3.10/site-packages/torch/serialization.py", line 1439, in find_class
    return super().find_class(mod_name, name)
ModuleNotFoundError: No module named 'torch.distributed.fsdp.flat_param'
```

It looks like `flat_param` was renamed to `_flat_param` at some point in Pytorch, which is what is causing this issue.

### Replication

1. Pull most recent version of OLMo repo
2. Create a conda env and `pip install -e .[all]`
3. After downloading a mitchish7 checkpoint (to the `$ROOT` path below), I run:

```shell
ROOT=${ROOT:-"/net/nfs.cirrascale/allennlp/willm/olmo-sparsity/checkpoints"}
ckpt="OLMo-medium/mitchish7/step0"

python scripts/unshard.py \
      $ROOT/$ckpt \
      $ROOT/$ckpt-unsharded \
      --type local \
      --safe-tensors
```

This produces the `ModuleNotFoundError` shown above.

### Versions

Python 3.10.9

<details>
<summary>pip freeze</summary>
accelerate==1.0.1<br />
-e git+https://github.com/allenai/OLMo.git@afd728f218cbec22fda20a254390c580de5bf4ae#egg=ai2_olmo<br />
ai2-olmo-core==0.1.0<br />
aiohappyeyeballs==2.4.3<br />
aiohttp==3.10.10<br />
aiosignal==1.3.1<br />
annotated-types==0.7.0<br />
antlr4-python3-runtime==4.9.3<br />
anyio==4.6.0<br />
argon2-cffi==23.1.0<br />
argon2-cffi-bindings==21.2.0<br />
arrow==1.3.0<br />
asttokens==2.4.1<br />
async-lru==2.0.4<br />
async-timeout==4.0.3<br />
attrs==24.2.0<br />
babel==2.16.0<br />
backoff==2.1.2<br />
backports.tarfile==1.2.0<br />
beaker-gantry==1.8.3<br />
beaker-py==1.32.0<br />
beautifulsoup4==4.12.3<br />
black==23.12.1<br />
bleach==6.1.0<br />
blessed==1.20.0<br />
boltons==24.0.0<br />
boto3==1.35.38<br />
boto3-extensions==0.23.0<br />
botocore==1.35.38<br />
brotlipy==0.7.0<br />
build==1.2.2.post1<br />
cached_path==1.6.3<br />
cachetools==5.5.0<br />
certifi @ file:///croot/certifi_1671487769961/work/certifi<br />
cffi @ file:///croot/cffi_1670423208954/work<br />
charset-normalizer @ file:///tmp/build/80754af9/charset-normalizer_1630003229654/work<br />
click==8.1.7<br />
click-aliases==1.0.5<br />
click-help-colors==0.9.4<br />
cmake==3.30.5<br />
colorama==0.4.6<br />
comm==0.2.2<br />
conda==23.1.0<br />
conda-content-trust @ file:///tmp/abs_5952f1c8-355c-4855-ad2e-538535021ba5h26t22e5/croots/recipe/conda-content-trust_1658126371814/work<br />
conda-package-handling @ file:///croot/conda-package-handling_1672865015732/work<br />
conda_package_streaming @ file:///croot/conda-package-streaming_1670508151586/work<br />
contourpy==1.3.0<br />
cryptography @ file:///croot/cryptography_1673298753778/work<br />
cycler==0.12.1<br />
datasets==2.7.1<br />
dateparser==1.2.0<br />
debugpy==1.8.6<br />
decorator==5.1.1<br />
defusedxml==0.7.1<br />
dill==0.3.6<br />
docker==7.1.0<br />
docker-pycreds==0.4.0<br />
docutils==0.21.2<br />
einops==0.8.0<br />
exceptiongroup==1.2.2<br />
executing==2.1.0<br />
face==20.1.1<br />
fastjsonschema==2.20.0<br />
filelock==3.13.4<br />
fire==0.7.0<br />
fonttools==4.54.1<br />
fqdn==1.5.1<br />
frozenlist==1.4.1<br />
fsspec==2024.6.1<br />
ftfy==6.3.0<br />
gitdb==4.0.11<br />
GitPython==3.1.43<br />
glom==23.5.0<br />
gluonts==0.15.1<br />
google-api-core==2.21.0<br />
google-auth==2.35.0<br />
google-cloud-core==2.4.1<br />
google-cloud-storage==2.18.2<br />
google-crc32c==1.6.0<br />
google-resumable-media==2.7.2<br />
googleapis-common-protos==1.65.0<br />
gpustat==1.1.1<br />
h11==0.14.0<br />
halo==0.0.31<br />
httpcore==1.0.6<br />
httpx==0.27.2<br />
huggingface-hub==0.23.5<br />
hydra-core==1.3.2<br />
idna @ file:///croot/idna_1666125576474/work<br />
importlib_metadata==8.5.0<br />
importlib_resources==6.4.5<br />
iniconfig==2.0.0<br />
ipykernel==6.29.5<br />
ipython==8.28.0<br />
ipywidgets==8.1.5<br />
isodate==0.7.2<br />
isoduration==20.11.0<br />
isort==5.12.0<br />
jaraco.classes==3.4.0<br />
jaraco.context==6.0.1<br />
jaraco.functools==4.1.0<br />
jedi==0.19.1<br />
jeepney==0.8.0<br />
Jinja2==3.1.4<br />
jmespath==1.0.1<br />
joblib==1.4.2<br />
json5==0.9.25<br />
jsonpointer==3.0.0<br />
jsonschema==4.23.0<br />
jsonschema-specifications==2023.12.1<br />
jupyter==1.1.1<br />
jupyter-console==6.6.3<br />
jupyter-events==0.10.0<br />
jupyter-lsp==2.2.5<br />
jupyter_client==8.6.3<br />
jupyter_core==5.7.2<br />
jupyter_server==2.14.2<br />
jupyter_server_terminals==0.5.3<br />
jupyterlab==4.2.5<br />
jupyterlab_pygments==0.3.0<br />
jupyterlab_server==2.27.3<br />
jupyterlab_widgets==3.0.13<br />
keyring==25.4.1<br />
kiwisolver==1.4.7<br />
lightning-utilities==0.11.7<br />
lit==18.1.8<br />
llvmlite==0.43.0<br />
log-symbols==0.0.14<br />
markdown-it-py==3.0.0<br />
MarkupSafe==2.1.5<br />
matplotlib==3.9.2<br />
matplotlib-inline==0.1.7<br />
mdurl==0.1.2<br />
mistune==3.0.2<br />
more-itertools==10.5.0<br />
mpmath==1.3.0<br />
msgspec==0.18.6<br />
multidict==6.1.0<br />
multiprocess==0.70.14<br />
mypy==1.3.0<br />
mypy-extensions==1.0.0<br />
nbclient==0.10.0<br />
nbconvert==7.16.4<br />
nbformat==5.10.4<br />
necessary==0.4.3<br />
nest-asyncio==1.6.0<br />
networkx==3.3<br />
nh3==0.2.18<br />
notebook==7.2.2<br />
notebook_shim==0.2.4<br />
numba==0.60.0<br />
numpy==1.26.4<br />
nvidia-cublas-cu12==12.1.3.1<br />
nvidia-cuda-cupti-cu12==12.1.105<br />
nvidia-cuda-nvrtc-cu12==12.1.105<br />
nvidia-cuda-runtime-cu12==12.1.105<br />
nvidia-cudnn-cu12==8.9.2.26<br />
nvidia-cufft-cu12==11.0.2.54<br />
nvidia-curand-cu12==10.3.2.106<br />
nvidia-cusolver-cu12==11.4.5.107<br />
nvidia-cusparse-cu12==12.1.0.106<br />
nvidia-ml-py==12.560.30<br />
nvidia-nccl-cu12==2.20.5<br />
nvidia-nvjitlink-cu12==12.6.68<br />
nvidia-nvtx-cu12==12.1.105<br />
omegaconf==2.3.0<br />
oocmap==0.3<br />
overrides==7.7.0<br />
packaging==24.1<br />
pandas==2.2.3<br />
pandocfilters==1.5.1<br />
parso==0.8.4<br />
pathspec==0.12.1<br />
petname==2.6<br />
pexpect==4.9.0<br />
pillow==10.4.0<br />
pkginfo==1.10.0<br />
platformdirs==4.3.6<br />
pluggy==1.5.0<br />
polars==1.9.0<br />
prometheus_client==0.21.0<br />
prompt_toolkit==3.0.48<br />
propcache==0.2.0<br />
proto-plus==1.24.0<br />
protobuf==5.28.2<br />
psutil==6.0.0<br />
ptyprocess==0.7.0<br />
pure_eval==0.2.3<br />
pyarrow==17.0.0<br />
pyasn1==0.6.1<br />
pyasn1_modules==0.4.1<br />
pycosat @ file:///croot/pycosat_1666805502580/work<br />
pycparser @ file:///tmp/build/80754af9/pycparser_1636541352034/work<br />
pydantic==2.9.2<br />
pydantic_core==2.23.4<br />
Pygments==2.18.0<br />
pyOpenSSL @ file:///opt/conda/conda-bld/pyopenssl_1643788558760/work<br />
pyparsing==3.1.4<br />
pyproject_hooks==1.2.0<br />
pyrootutils==1.0.4<br />
PySocks @ file:///home/builder/ci_310/pysocks_1640793678128/work<br />
pytest==8.3.3<br />
pytest-sphinx==0.6.3<br />
python-dateutil==2.9.0.post0<br />
python-dotenv==1.0.1<br />
python-json-logger==2.0.7<br />
pytorch-lightning==2.0.4<br />
pytz==2024.2<br />
PyYAML==6.0.2<br />
pyzmq==26.2.0<br />
readme_renderer==44.0<br />
referencing==0.35.1<br />
regex==2024.9.11<br />
requests==2.32.3<br />
requests-toolbelt==1.0.0<br />
requirements-parser==0.11.0<br />
responses==0.18.0<br />
rfc3339-validator==0.1.4<br />
rfc3986==2.0.0<br />
rfc3986-validator==0.1.1<br />
rich==13.9.1<br />
rpds-py==0.20.0<br />
rsa==4.9<br />
ruamel.yaml @ file:///croot/ruamel.yaml_1666304550667/work<br />
ruamel.yaml.clib @ file:///croot/ruamel.yaml.clib_1666302247304/work<br />
ruff==0.6.9<br />
s3transfer==0.10.3<br />
s4 @ git+https://github.com/jopetty/s4.git@8a30805b0bea14082a133f59342b78acc7c23f23<br />
safetensors==0.4.5<br />
scikit-learn==1.5.2<br />
scipy==1.14.1<br />
SecretStorage==3.3.3<br />
Send2Trash==1.8.3<br />
sentry-sdk==2.16.0<br />
setproctitle==1.3.3<br />
sfirah @ git+https://github.com/jopetty/sfirah@30b5a87342882f42dc3ed4136ee1ade3633c266e<br />
six @ file:///tmp/build/80754af9/six_1644875935023/work<br />
smart-open==7.0.5<br />
smashed==0.21.5<br />
smmap==5.0.1<br />
sniffio==1.3.1<br />
soupsieve==2.6<br />
spinners==0.0.24<br />
stack-data==0.6.3<br />
sympy==1.13.3<br />
tabulate==0.9.0<br />
termcolor==2.5.0<br />
terminado==0.18.1<br />
threadpoolctl==3.5.0<br />
timm==0.5.4<br />
tinycss2==1.3.0<br />
tokenizers==0.20.1<br />
tomli==2.0.2<br />
toolz @ file:///croot/toolz_1667464077321/work<br />
torch==2.3.1<br />
torchmetrics==1.4.3<br />
torchtext==0.18.0<br />
torchvision==0.18.1<br />
tornado==6.4.1<br />
tqdm==4.66.5<br />
traitlets==5.14.3<br />
transformers==4.45.2<br />
triton==2.3.1<br />
trouting==0.3.3<br />
twine==5.1.1<br />
typeguard==2.13.3<br />
types-python-dateutil==2.9.0.20241003<br />
types-setuptools==75.1.0.20240917<br />
typing_extensions==4.12.2<br />
tzdata==2024.2<br />
tzlocal==5.2<br />
uri-template==1.3.0<br />
urllib3 @ file:///croot/urllib3_1673575502006/work<br />
wandb==0.18.3<br />
wcwidth==0.2.13<br />
webcolors==24.8.0<br />
webencodings==0.5.1<br />
websocket-client==1.8.0<br />
widgetsnbextension==4.0.13<br />
wrapt==1.16.0<br />
xxhash==3.5.0<br />
yarl==1.14.0<br />
zipp==3.20.2<br />
zstandard @ file:///opt/conda/conda-bld/zstandard_1663827383994/work<br />

</details>

**Comment by viking-sudo-rm** [comment]
Update: upgrading to pytorch 2.4 did not fix the issue!

I ran `pip install --upgrade torch==2.4` and then reran src/unshard.py, which produced the same error. I added a breakpoint to the unshard script and verified the version had been updated:
```python
-> model_state_dict, optim_state_dict, trainer_state_dict = checkpointer.unshard_checkpoint(
(Pdb) torch.__version__
'2.4.0+cu121'
```

**Comment by viking-sudo-rm** [comment]
Issue was solved by downgrading to torch 2.1:
```shell
pip install --upgrade torch==2.1
```
